### PR TITLE
feat: changes in schema, refactorings, input loops, random parameters in channels.

### LINKED ISSUE #48: Simulating user-flows
#### Reason/Context
I would love to be able to define user flows in a scenario file.

#### Description
Let's say I want to simulate specific user flows in a user-registration service. It could maybe look something along the lines of:
```
user-logs-on:
user-logs-on-when-logedin:
user-logs-out:
```

This would be a breaking change to the scenario spec.

### PR DESCRIPTION
**Pull request by NektariosFifes**
**Description**

- Major improvements in schema design.
-  Related changes to make these changes work.
- Added random generator of topics.
- Changes in api. Now two main methods exposed are startOperation() and startScenario().

**Related issue(s)**
fixes #48

### CODE REVIEW

**Comment by NektariosFifes** [comment]
@jonaslagoni just to inform u most of it done. there are some eslint errors i need to fix but structurally it is done. if u want take a look.

**Comment by NektariosFifes** [comment]
> @jonaslagoni just to inform u most of it done. there are some eslint errors i need to fix but structurally it is done. if u want take a look.

actually nevermind its not ready yet it has some issue with the parameters replacement in the channel string

**Comment by NektariosFifes** [comment]
@jonaslagoni  ok it is ready :)

**Comment by jonaslagoni** [comment]
@NektariosFifes you have multiple security issues and code smells that need to be addressed before I can review 🙂

**Comment by NektariosFifes** [comment]
@jonaslagoni I fixed the first code smell. The rest of them cannot be fixed. They require refactoring of the entire project, and too much abstraction. Abstraction is nice but it does not make sense to use it here as i would need 5-6 new functions in the mqtt handler. Also he has some issue with me using (?question mark if). I think it makes my code much prettier.
Can we ignore these and merge it cause i need it for the desktop app to function. :)

File: example-projects/game-processor/AsyncAPI/components/messages/PlayerItemPickup.yaml [hunk 1]
```diff
@@ -1,9 +0,0 @@
-payload:
-  type: object
```
Comment (jonaslagoni) [inline example-projects/game-processor/AsyncAPI/components/messages/PlayerItemPickup.yaml:2]: Why is this removed?

File: example-projects/game-processor/index.js [hunk 2]
```diff
@@ -3,7 +3,7 @@ const mqtt = require('async-mqtt');
 const { v4: uuidv4 } = require('uuid');
 const MongoClient = require('mongodb').MongoClient;
 const Config = require('./config');
-var mongodbUrl = `mongodb://${Config.mongodbUsername}:${Config.mongodbPassword}@${Config.mongodbHost}`;
+var mongodbUrl = `mongodb://localhost:27017/processor`;
```
Comment (jonaslagoni) [inline example-projects/game-processor/index.js:6]: Why change this from using configs?

File: example-projects/game-processor/scenario.yaml [hunk 3]
```diff
@@ -1,11 +1,42 @@
...
+      max: 2000
+    serverId:
+      min: 0
+      max: 4
+user-gameLoop:
+  loop:
+    interval:
+      600
+    cycles:
+      5
```
Comment (jonaslagoni) [inline example-projects/game-processor/scenario.yaml:15]: Feel free to create multiple scenarios, so you don't change existing useable scenarios 🙂

File: example-projects/game-processor/scenario.yaml [hunk 4]
```diff
@@ -1,11 +1,42 @@
...
+    game/server/{serverId}/events/player/{playerId}/hit:
+      serverId: '1'
+      playerId:
+        regex: '^[\w\d]{1,22}$'
+      payload:
+        crit: 125
+        apDamage: 30
```
Comment (jonaslagoni) [inline example-projects/game-processor/scenario.yaml:22]: Why are you generating payloads that are incompatible with the specified message payload structure? 🤔 

https://github.com/asyncapi/simulator/blob/dc8d236869b619347cc312035462a27f73ca47e3/example-projects/game-processor/AsyncAPI/components/messages/PlayerHit.yaml

File: src/RequestHandler/mqttHandler.js [hunk 5]
```diff
@@ -1,36 +1,133 @@
...
+          currentCycle += 1;
+          console.log(channelName);
+          await client.publish(channelName, JSON.stringify(payload));
+          if (currentCycle === cycles)
+            clearInterval(actionLoop);
+        }, interval);
+
+        aliveOperations[operatioName] = {
+          loopInstance: actionLoop
+        };
```
Comment (jonaslagoni) [inline src/RequestHandler/mqttHandler.js:75]: This could be contained in something like `runOperationForChannel` to reduce complexity.

File: src/Schema/SimulationPlotSchema_0.0.1.json [hunk 6]
```diff
@@ -4,52 +4,145 @@
...
-          "description": "Events per second being emitted",
-          "type": "integer"
-        },
-        "priority" : {
-          "type": "integer",
-          "description": "This determines whether it should be prioritized in case the resources to emit events is insufficient."
-        }
+    "^scenario-[\\w\\d]+$" : {
+      "type": "array",
+      "items": {
```
Comment (jonaslagoni) [inline src/Schema/SimulationPlotSchema_0.0.1.json:9]: I would highly recommend start splitting up the schema file into digestible chunks, cause these are really quickly becoming unreadable 🙂

File: src/parser/GenerateOperationsAndScenarios.js [hunk 7]
```diff
@@ -0,0 +1,99 @@
+// eslint-disable-next-line sonarjs/cognitive-complexity
+function getdefinedChannels (channels) {
```
Comment (jonaslagoni) [inline src/parser/GenerateOperationsAndScenarios.js:2]: `getdefinedChannels` should probably be called `getDefinedChannels´.

File: src/parser/GenerateOperationsAndScenarios.js [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+      if (value._json['x-plot']) {
+        subscribeChannels.soloOps[value._json['x-plot']] = Object.assign({}, value.subscribe()._json, {route: key});
+      }
+      if (value._json['x-group']) {
+        if (!subscribeChannels.groupOps[value._json['x-group']]) {
+          subscribeChannels.groupOps[value._json['x-group']] = {};
+        }
+        Object.assign(subscribeChannels.groupOps[value._json['x-group']], {[key]: value.subscribe()._json});
+      }
+    }
```
Comment (jonaslagoni) [inline src/parser/GenerateOperationsAndScenarios.js:37]: This could be grouped in a `getDefinedChannel` to reduce complexity.

File: src/parser/GenerateOperationsAndScenarios.js [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+
+  return operations;
+}
+
+function generateScenarios(scenarioParsed,operations) {
+  const scenarios = {};
+  for (const [itemName,value] of Object.entries(scenarioParsed)) {
+    if (itemName.match(RegExp(/^scenario-[\w\d]+$/))) {
+      // eslint-disable-next-line security/detect-object-injection
+      scenarios[itemName]  = {};
```
Comment (jonaslagoni) [inline src/parser/GenerateOperationsAndScenarios.js:59]: Please address the issue rather than disable it. 

```suggestion
      scenarios[string(itemName)]  = {};
```

File: src/parser/GenerateOperationsAndScenarios.js [hunk 10]
```diff
@@ -0,0 +1,99 @@
...
+      for (const operationName of value) {
+        for (const [name,value] of Object.entries(operations)) {
+          if (name === operationName) {
+            // eslint-disable-next-line security/detect-object-injection
```
Comment (jonaslagoni) [inline src/parser/GenerateOperationsAndScenarios.js:63]: Same here, address the issue rather then disabling the check, otherwise why have it 😄

File: src/parser/tests/parser.test.js [hunk 11]
```diff
@@ -1,17 +1,19 @@
...
+  // eslint-disable-next-line no-unused-vars
   let asyncApi;
   it('Should be able to parse correctly formatted AsyncAPI and scenario files',async function() {
     try {
       [asyncApi] = await parseFiles(path.resolve(__dirname, './files/correct/CorrectlyFormattedAsyncApi.yaml'), path.resolve(__dirname, './files/correct/CorrectlyFormattedScenario.yaml'));
     } catch (err) {
       assert.fail('Test failed. Parser was not able to parse one of the files you provided');
     }
-    expect(asyncApi._json).to.deep.include(expectedOutputs.p1AsyncApi._json);
+    //expect(asyncApi._json).to.equal(expectedOutputs.p1AsyncApi._json);
```
Comment (jonaslagoni) [inline src/parser/tests/parser.test.js:16]: Why disable the check?

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
Have a few questions and suggestions.

Going forward, try not to mix refactoring and features together. It gets way too complicated to review 😅 

> The rest of them cannot be fixed. They require refactoring of the entire project, and too much abstraction. Abstraction is nice but it does not make sense to use it here as i would need 5-6 new functions in the mqtt handler.

The reason the code should be split up is that it is good practice. It makes the code chunks easier to read for others. You might understand what is happening, others definitely dont and it would take longer for them to understand the code when everything is added under the same function 🙂

I have added a few suggestions on how it can be done.

File: src/parser/tests/parser.test.js (hunk 11 above)
Comment (NektariosFifes) [reply src/parser/tests/parser.test.js:16]: it is this weird thing with expect. The object are 100% equal but still it will throw an error. It is the same thing we discussed during gsoc.

File: example-projects/game-processor/index.js (hunk 2 above)
Comment (NektariosFifes) [reply example-projects/game-processor/index.js:6]: There is an issue here usually when sm1 install mongodb he doesnt not have a username or password so the example will not work in some cases. i will try to make sure it works without setting a user and password

File: src/Schema/SimulationPlotSchema_0.0.1.json (hunk 6 above)
Comment (NektariosFifes) [reply src/Schema/SimulationPlotSchema_0.0.1.json:9]: i will split it with subschemas in the same file. there is some weird issue with json schema, it is not able to resolve external schemas.

File: example-projects/game-processor/AsyncAPI/components/messages/PlayerItemPickup.yaml (hunk 1 above)
Comment (NektariosFifes) [reply example-projects/game-processor/AsyncAPI/components/messages/PlayerItemPickup.yaml:2]: by accident i will add it.

File: package.json [hunk 12]
```diff
@@ -3,7 +3,7 @@
   "version": "0.4.7",
   "type": "commonjs",
   "description": "The asyncapi command line library to test your app at scale.",
-  "main": "src/bin/cli.js",
```
[the remark this answers is not in the document]
Comment (NektariosFifes) [reply package.json:6]: the problem is that when i import a module from the library for example the parser, (for usage in the desktop application) the main script will run and stop the application (cause it needs arguments).

[the remark this answers is not in the document]
Comment (jonaslagoni) [reply package.json:6]: 🤔 Is there a reason to have the CLI code present in the repository?

File: src/parser/tests/parser.test.js [hunk 13]
```diff
@@ -1,17 +1,20 @@
 const assert = require('assert');
 const expect = require('chai').expect;
 const path = require('path');
+// eslint-disable-next-line no-unused-vars
```
Comment (jonaslagoni) [inline src/parser/tests/parser.test.js:4]: This can be removed right? 🤔

File: src/parser/tests/parser.test.js [hunk 14]
```diff
@@ -1,17 +1,20 @@
...
 const expectedOutputs = require('./expectedOutputs');
+const nodeAssert = require('assert');
 const {parseFiles} = require('../parseFiles');
 describe('Parser',function() {
+  // eslint-disable-next-line no-unused-vars
```
Comment (jonaslagoni) [inline src/parser/tests/parser.test.js:9]: This can be removed right? 🤔

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
@NektariosFifes it is hard for me to know if the previous comments have been addressed if they are not `resolved`, make sure to do that before re-requesting review 🙂

File: package.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (NektariosFifes) [reply package.json:6]: you think i should move it to the cli project?

File: src/parser/tests/parser.test.js (hunk 14 above)
Comment (NektariosFifes) [reply src/parser/tests/parser.test.js]: yes

File: example-projects/game-processor/scenario.yaml (hunk 3 above)
Comment (NektariosFifes) [reply example-projects/game-processor/scenario.yaml:15]: i will try to do it in a future pull request

File: package.json (hunk 12 above)
[the remark this answers is not in the document]
Comment (NektariosFifes) [reply package.json:6]: i think i will keep it here for now and probably move it later on

**Comment by NektariosFifes** [comment]
@jonaslagoni i removed the eslint comments :)

File: src/RequestHandler/mqttHandler.js [hunk 15]
```diff
@@ -1,36 +1,144 @@
...
+  return new randExp(regex).gen();
+}
+
+function getChannelParams (channel) {
+  return channel.match(new RegExp(/{(.*?)}/gm)).map((item) => item.substring(1, item.length - 1));
+}
+/* eslint-disable security/detect-object-injection */
+/* eslint-disable no-unused-expressions */
+/* eslint-disable prefer-const */
+// eslint-disable-next-line sonarjs/cognitive-complexity
```
Comment (jonaslagoni) [inline src/RequestHandler/mqttHandler.js:59]: What is the reason for disabling all the rules?

File: src/RequestHandler/mqttHandler.js [hunk 16]
```diff
@@ -1,36 +1,144 @@
...
+    if (operationData.loop) {
+      const channels = Object.assign({},operationData.loop);
+
+      const cycles = operationData.loop.cycles;
+      const interval = operationData.loop.interval;
+
+      delete channels.interval;
+      delete channels.cycles;
+
+      // eslint-disable-next-line prefer-const
```
Comment (jonaslagoni) [inline src/RequestHandler/mqttHandler.js:84]: Why disable this?

File: src/Schema/SimulationPlotSchema_0.0.1.json [hunk 17]
```diff
@@ -4,59 +4,164 @@
...
+      ]
+    }
+  },
+  "required": ["version"],
+
+  "definitions": {
+    "scenarioSchema" : {
+      "title": "scenarioSchema.json",
+      "type": "array",
+      "definitions": {},
```
Comment (jonaslagoni) [inline src/Schema/SimulationPlotSchema_0.0.1.json:29]: Why this empty definition? 🤔

File: src/Schema/SimulationPlotSchema_0.0.1.json [hunk 18]
```diff
@@ -4,59 +4,164 @@
...
+                        }
+                      }
+                    ]
+                  }
+                }
+              }
+            }
+          }
+        ]
       }
```
Comment (jonaslagoni) [inline src/Schema/SimulationPlotSchema_0.0.1.json:100]: IMO each object here, should be split into their own definition, easier to read 🙂

File: src/Schema/SimulationPlotSchema_0.0.1.json [hunk 19]
```diff
@@ -4,59 +4,164 @@
...
+                          "type": "integer"
+                        },
+                        "max": {
+                          "type": "integer"
+                        }
+                      }
+                    }
+                  }
+                }
+              ]
```
Comment (jonaslagoni) [inline src/Schema/SimulationPlotSchema_0.0.1.json:160]: Same here, I guess you can reuse same definition as above?

File: src/bin/cli.js [hunk 20]
```diff
@@ -172,10 +172,10 @@ const cli = async () => {
     asyncApiPath = path.resolve(options.filepath);
     scenarioPath = path.resolve(options.scenario);
   }
-  const structuredData = await verifyInputGetData(rdInterface, path.resolve(asyncApiPath),path.resolve(scenarioPath),options.basedir);
+  const dataFromParser = await verifyInputGetData(rdInterface, path.resolve(asyncApiPath),path.resolve(scenarioPath),options.basedir);
   const manager = RequestManager();
-  await manager.createReqHandler(structuredData);
-  await manager.startOperations();
+  await manager.createReqHandler(dataFromParser);
+  await manager.startScenario('SimpleGame','mqtt');
```
Comment (jonaslagoni) [inline src/bin/cli.js:178]: 🧐 

Why is this hardcoded in the CLI? 

Does this mean that you cannot use the CLI for other scenarios?

**Review by jonaslagoni — CHANGES_REQUESTED** [review]
A few comments left 👍

File: src/bin/cli.js (hunk 20 above)
Comment (NektariosFifes) [reply src/bin/cli.js:178]: i will change it to execute all scenarios

Comment (NektariosFifes) [reply src/bin/cli.js:178]: i think i will limit the cli to executing all the scenarios. For more specific things sm1 would have to use the javascript api, or the desktop application. At least for now

File: src/Schema/SimulationPlotSchema_0.0.1.json (hunk 17 above)
Comment (NektariosFifes) [reply src/Schema/SimulationPlotSchema_0.0.1.json:29]: i removed it

File: src/RequestHandler/mqttHandler.js (hunk 16 above)
Comment (NektariosFifes) [reply src/RequestHandler/mqttHandler.js:84]: i removed it

File: src/RequestHandler/mqttHandler.js (hunk 15 above)
Comment (NektariosFifes) [reply src/RequestHandler/mqttHandler.js:59]: i removed the rules and moved the disable prefer-const where it should be. Did not delete it cause it seems that eslint has an issue with me using shorthand syntax to declare variables to get enumerator values.
for(let [name,value] of Object.entries(channelParams)) . value is reasigned and name is not. sadly i cannot declare one to be let and other to be constant. Also eslint-disable-next-line sonarjs/cognitive-complexity cannot be removed. Me using a function as context in other to return a interface of functions binded to that context is a programming paradigm in javascript. I dont think it is classified as an error.

**Review by jonaslagoni — APPROVED** [review]
LGTM 👍 

Going forward, please don't group so many different changes inside one PR. It is a nightmare to review. 

Ideally, this PR should have been split up into:
1. Changes to the schema.
2. Refactoring input loops
3. Adding random values

This way, reviewing it wont take ages to do either 🙂 Just keep it in mind!

**Comment by NektariosFifes** [comment]
@derberg these sonarcloud issues are pretty overkill. i mean denial of service??? can we bypass them?

**Comment by jonaslagoni** [comment]
> @derberg these sonarcloud issues are pretty overkill. i mean denial of service??? can we bypass them?

Depending on the usage of the library, especially since you are building a platform, these issues become very relevant. If these issues remain, one bad user could deny service for others 🙂

**Comment by NektariosFifes** [comment]
@jonaslagoni i fixed the security issues by using a timeout inside a promise that returns the regex result. that way the regex cannot get too large and crash the app. These sonarcloud issues seem to need a human review , seems like they wont solve automatically.

**Comment by NektariosFifes** [comment]
also the pseudonumber issue is not important here. we dont really care about the random number being guessed.

**Review by jonaslagoni — APPROVED** [review]

**Comment by jonaslagoni** [comment]
@derberg how do we make sonarcloud ignore the specific findings so this can be merged?

**Comment by derberg** [comment]
@jonaslagoni it is all done through the configuration in `.sonarcloud.properties` or inline as comments, afaik you add `//NOSONAR` at the end of the line that is causing issues. I only suggest that every single exclusion is properly commented for future maintainers

**Comment by NektariosFifes** [comment]
@derberg i cant find that file where is it located?

**Comment by NektariosFifes** [comment]
@derberg i disabled the checks for security that sonarcloud did not manage to realise are resolved. Yet it still bothers me about one section that it considers as code duplication. It is not code duplication and i cannot change it. I still dont understand what the usage of this sonarcloud tool is. Seems pretty annoying to me. No documentation whatsoever for disabling duplication checks.
it is 3.1% duplicaton and the limit is 3% . However i cannot reduce more. And i cant find anything on how to diable duplication checks.

**Comment by NektariosFifes** [comment]
Kinda biased opinion due to my frustration:
 sonarcloud is trash

**Comment by NektariosFifes** [comment]
can we remove it. It has 0 usage and only serves to annoy and slow down progress

**Comment by jonaslagoni** [comment]
Yea, I have had my fair share of clashes with sonarcloud. Something that just can't be denied, is that it forces one to develop more maintainable and easy-to-understand code. It is a bit of a struggle to get started with the tool I give you that 😄 

To ignore duplicate code you can use the following configuration file that we have in Modelina: https://github.com/asyncapi/modelina/blob/master/.sonarcloud.properties

I actually don't know if it can or should be disabled, @derberg are your thoughts?

**Comment by NektariosFifes** [comment]
did just sonarcloud approve my efforts?! :octocat:

**Review by jonaslagoni — APPROVED** [review]

**Comment by derberg** [comment]
Hey @NektariosFifes and @jonaslagoni 

So yeah, my personal statement is:
- tools like this are helpful as they from time to time find out bugs and let you also think about your code. I would for example never merge a PR with 8 code smells -> https://sonarcloud.io/project/issues?id=asyncapi_simulator&pullRequest=50&resolved=false&types=CODE_SMELL. For example 67 lines of code in one function seems like a lot, and there are only few exceptions where I could live in my life with 20-25 lines, but because further refactor would just make it harder to read.
- yes, it is not the nicest tool from the documentation point of view. I don't complain as we got it for free. There are 2 options if I find something frustrating because I have challenges with configuration, I document it or switch tool. So I would suggest you start working on a developer document that describes all hacks for sonarcloud so in the future other maintainers can benefit. Or take some time and research an alternative solution 😄 

From AsyncAPI Initiative point of view, we do not set any official requirements. We have ambitions that production-ready solutions should pass [this badge](https://bestpractices.coreinfrastructure.org/en/criteria/0). This is not required yet. Maybe will be in 2022 maybe not 🤷🏼 but is not for now.

So yeah, if you do not want to regularly check with SonarCloud but prefer to work on it in few months (or a year) to fix all the issues with SonarCloud or some other tool - up to you folks, you are code owners here, you are the ones responsible for the quality here. Just let me know and I can disable SonarCloud for this repository.
